[IL] HDX-10019: refactoring to async handlers;

odemkovych opened the pull request.

odemkovych committed:
[IL] HDX-10019: refactoring to async handlers;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

aggarwaa commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -85,10 +81,101 @@ class Table:
         "pip-system-certs",
     ],
-    auth=HydrolixCredentialChain(f"https://{HYDROLIX_CONFIG.host}/config"),
+    # auth=HydrolixCredentialChain(f"https://{HYDROLIX_CONFIG.host}/config"),
```
is this commented out for testing or some other reason?

odemkovych replied:
Yes, it was for tests. I'll revert. Thank you!

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

aggarwaa commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -89,6 +85,97 @@ class Table:
+                rows = []
+                for block in stream:
+                    rows.extend(block)
+                    asyncio.sleep(0.01)
```
does this need to be `await`ed? (why is the `sleep` needed in the first place?)

odemkovych replied:
Initially I was going to use 'await client.query`. But during loadtest iterations I experienced situation when service was unresponsive (rps=0) and it looked like the long queries did excessive load. I tried to get into it and decided to offload receiving onto main event loop. That behaved a bit better but obviously needed more testing. 
Thank you for pinpointing that part, I've continued with it today and tried to reproduce but all of no avail. That must be an issue with the test or my CH container.

aggarwaa commented on mcp_hydrolix/mcp_env.py:
```diff
@@ -163,6 +194,60 @@ def mcp_bind_port(self) -> int:
+        Only used when transport is "http" or "sse".
+        Default: 10000
+        """
+        return int(os.getenv("HYDROLIX_MCP_MAX_REQUESTS", 3000))
```
s/`HYDROLIX_MCP_MAX_REQUESTS`/`HYDROLIX_MCP_MAX_REQUESTS_JITTER`

aggarwaa commented on mcp_hydrolix/mcp_env.py:
```diff
@@ -163,6 +194,60 @@ def mcp_bind_port(self) -> int:
+        Only used when transport is "http" or "sse".
+        Default: 1000
+        """
+        return int(os.getenv("HYDROLIX_MCP_MAX_REQUESTS_JITTER", 500))
```
s/`HYDROLIX_MCP_MAX_REQUESTS_JITTER`/`HYDROLIX_MCP_MAX_REQUESTS`

aggarwaa commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -89,6 +85,97 @@ class Table:
+            with await client.query_row_block_stream(
+                query,
+                settings={
+                    "readonly": 0,
```
we want `readonly` to be `1` by default:
> "readonly": 1

odemkovych replied:
Next query "SELECT * FROM loop  (numbers(3)) LIMIT 300000" is rejected for execution when readonly=1.
That might be due to some temporary tables creation.

odemkovych replied:
Reverted.

aggarwaa commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -97,8 +184,8 @@ async def health_check(request: Request) -> PlainTextResponse:
         # Try to create a client connection to verify query-head connectivity
-        client = create_hydrolix_client(get_request_credential())
-        version = client.server_version
+        client = await create_hydrolix_client(client_shared_pool, get_request_credential())
```
will this leak the connection? the following might not:
> async with await create_hydrolix_client(client_shared_pool,
  get_request_credential()) as client

odemkovych replied:
Adding `with...`

btw: I've taken a look into the code a bit. It shouldn't create leak because `server_version` is initialized during the client creation by `command` call and it release connection in scope of it https://github.com/urllib3/urllib3/blob/720f484b605f18887a48eef448d0084e2b76902d/src/urllib3/connectionpool.py#L860

aggarwaa commented on mcp_hydrolix/mcp_env.py:
```diff
@@ -150,9 +181,9 @@ def mcp_bind_host(self) -> str:
 
         Only used when transport is "http" or "sse".
-        Default: "127.0.0.1"
+        Default: "0.0.0.0"
```
what's the benefit of listening on all interfaces?

odemkovych replied:
As I recall '127.0.0.1' interface accepts incoming connection from local host only. Please let me know if I should revert it.

aggarwaa replied:
If this is canonical for services behind traefik, i'm good with this change.

emanb29 replied:
it's a nit, but I'd prefer we leave 127.0.0.1 in the implementation and just override it in the dockerfile -- As a rule of thumb, I prefer hardcoded values to be the most secure option and only override them with weaker values in the contexts for which the weaker values are necessary

aggarwaa commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -89,6 +85,97 @@ class Table:
+async def execute_query(query: str):
+    try:
+        async with await create_hydrolix_client(
+            client_shared_pool, get_request_credential()
```
it might be useful to test the per request credentials use with multiple workers, if it hasn't already been done

odemkovych replied:
I've tested it with multiple workers. In jira comment https://hydrolix.atlassian.net/browse/HDX-10019?focusedCommentId=59735 where 'workers=3' are results with multiple worker processes.

aggarwaa reviewed (commented):
great work! i'll leave it to @emanb29 to approve.

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

odemkovych committed:
[IL] HDX-10019: refactor credential propagation; improve CH client creation with shared http client pool; refactored unittests;

emanb29 commented on .github/workflows/tests.yaml:
```diff
@@ -0,0 +1,55 @@
+          uv run pytest
+      - name: Stop Docker Compose services
+        if: always()
+        run: docker compose down
```
Make sure to install and run `pre-commit`: you can catch up on the commits that are already pushed with `pre-commit run --all-files`

emanb29 commented on ruff.toml:
Adding our own `ruff` config (while I do prefer the defaults set by this file to those in use by the repo currently) adds yet another layer of difficulty when trying to merge in changes from upstream. Are we at the point where we're committing to 'no more pulling changes from upstream'?

odemkovych replied:
Understood. I'll remove it. We always can add it when we need it.

emanb29 commented on ruff.toml:
```diff
@@ -0,0 +1,46 @@
+
+[lint.isort]
+combine-as-imports = true
+known-first-party = ["ariadne_core", "ariadne_pythia"]
```
remove references to internal projects (this is a public repo)

emanb29 commented on tests/test_tool.py:
```diff
@@ -1,102 +1,112 @@
-        self.assertEqual(columns["id"]["comment"], "Primary identifier")
-        self.assertEqual(columns["name"]["comment"], "User name field")
+        self.assertEqual(columns["id"].comment, "Primary identifier")
+        self.assertEqual(columns["name"].comment, "User name field")
```
Similar to my comment on ruff.toml -- While I do prefer the style you've converted this to, it will introduce effectively no-op changes that make it hard to merge in from upstream. We should be deliberate in deciding whether that's the direction we want to take.

odemkovych replied:
I'll keep it the original style.

emanb29 commented on mcp_hydrolix/utilities.py:
```diff
@@ -0,0 +1,52 @@
+
+
+def with_serializer(fn):
+    """Decorator to apply custom serialization to tool output."""
```
document expectations of wrapped function (eg, "returns a ToolResult")

emanb29 commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -89,6 +84,95 @@ class Table:
+signal.signal(signal.SIGQUIT, term)
+
+
+async def execute_query(query: str):
```
nit: Annotate this, at least with TypedDict

emanb29 commented on mcp_hydrolix/log/log.py:
```diff
@@ -0,0 +1,60 @@
+import yaml
+
+
+class JsonFormatter(logging.Formatter):
```
Does this maintain redaction of the `token` parameter via `AccessLogTokenRedactingFilter`?

emanb29 commented on mcp_hydrolix/log/utils.py:
nit: move in to the `log` module

emanb29 commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -89,6 +84,95 @@ class Table:
+# allow custom hydrolix settings in CH client
+common.set_setting("invalid_setting_action", "send")
+common.set_setting("autogenerate_session_id", False)
+client_shared_pool = httputil.get_pool_manager(maxsize=get_config().query_pool_size, num_pools=1)
```
```suggestion
client_shared_pool = httputil.get_pool_manager(maxsize=HYDROLIX_CONFIG.query_pool_size, num_pools=1)
```

emanb29 commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -89,6 +84,95 @@ class Table:
+                query,
+                settings={
+                    "readonly": 1,
+                    "hdx_query_max_execution_time": get_config().query_timeout_sec,
```
```suggestion
                    "hdx_query_max_execution_time": HYDROLIX_CONFIG.query_timeout_sec,
```

emanb29 reviewed (changes requested):
Some minor things, and one project direction question we'll probably need to take offline. The 'heart' of the change (switching to the async variant of clickhouse client, including HTTP connection pool management) looks good.

odemkovych committed:
Update mcp_hydrolix/mcp_server.py

Co-authored-by: Ethan Bell <ebell@hydrolix.io>

odemkovych committed:
[IL] HDX-10019: code review changes;

odemkovych committed:
[IL] HDX-10019: formatting;

odemkovych committed:
[IL] HDX-10019: workflow;

odemkovych committed:
[IL] HDX-10019: unittests fixes;

odemkovych committed:
[IL] HDX-10019: unittests patch 'readonly' flag;

odemkovych committed:
[IL] HDX-10019: upgrade dependencies; remove cache action because UV action triggers caching as well;

odemkovych commented:
I've reviewed and refactored the remarks above.
Some details:
- Tests for the logs integration with the filter https://github.com/hydrolix/mcp-hydrolix/blob/87bd75371894d8aae80fd67200109e0c398b5737/tests/test_log.py#L250
- Had to keep TypedDict here for proper json serialisation https://github.com/hydrolix/mcp-hydrolix/blob/4b732721dd2be030ad2b81bccf7ee7ad300d3e8b/mcp_hydrolix/mcp_server.py#L66
- Added '__dict__' here for the original dictionary access pattern. I can add TypedDict to the involved model class but not sure it is a good call. Perhaps I will have some better idea later. https://github.com/hydrolix/mcp-hydrolix/blob/d33d63b7426d9487ff38b6b656330ac6d756280f/tests/test_tool.py#L104

emanb29 commented:
Sorry for the whiplash here -- FWIW, based on the conversation in Slack between Alok and I, we're both on board with whatever ruff / style changes you find best (I've really liked the ruff rules and style changes you applied). Either the 'low diff' style or the 'our own best practices' style is totally fine for this PR, and I don't consider it a blocker either way.

emanb29 commented on mcp_hydrolix/mcp_server.py:
```diff
@@ -43,37 +47,34 @@ class Table:
 
 # Add token redaction filter to uvicorn access logs
 uvicorn_access_logger = logging.getLogger("uvicorn.access")
 uvicorn_access_logger.addFilter(AccessLogTokenRedactingFilter())
```
can/should/is this be configured via `log.yaml` instead? (I didn't initially because TBH I'm not familiar enough with the python logging library)

odemkovych replied:
It is configured via log.yaml. I've removed this code block and tested manually.

emanb29 reviewed (approved):
Bangin' 🥳

odemkovych committed:
[IL] HDX-10019: remove logger configuration because it is configured via logging;
